tests/src/tests/test_utils.go

```diff
@@ -0,0 +1,394 @@
+	setRandSeed()
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = schemaBytes[rand.Intn(len(schemaBytes)-1)]
+	}
+	randomString := string(b)
+	if lower {
+		randomString = strings.ToLower(randomString)
+	}
+	return randomString
+}
+
+var GroupKinds = [...]schema.CommandGroupKind{schema.BuildCommandGroupKind, schema.RunCommandGroupKind, schema.TestCommandGroupKind, schema.DebugCommandGroupKind}
+
+// Return random group kind. One of "build", "run", "test" or "debug"
+func GetRandomGroupKind() schema.CommandGroupKind {
+	return GroupKinds[GetRandomNumber(len(GroupKinds))-1]
+}
+
+// Randomly returns true or false
+func GetBinaryDecision() bool {
+	return GetRandomDecision(1, 1)
+}
+
+// Randomly returns true or false, but weighted to one or the other.
+// For example if success is set to 2 and failure to 1, true is twice as likely to be returned.
+func GetRandomDecision(success int, failure int) bool {
+	setRandSeed()
+	return rand.Intn(success+failure) > failure-1
+}
+
+// Randomly returns an integer between 1 and the number specified.
+func GetRandomNumber(max int) int {
+	setRandSeed()
+	return rand.Intn(max) + 1
+}
+
+// Return a structure used to represent a specific devfile in the tests
+func GetDevfile(fileName string) TestDevfile {
+	testDevfile := TestDevfile{}
+	testDevfile.SchemaDevFile = schema.Devfile{}
+	testDevfile.FileName = fileName
+	testDevfile.SchemaDevFile.SchemaVersion = "2.0.0"
+	testDevfile.SchemaParsed = false
+	return testDevfile
+}
+
+// Create a devifle on disk for use in the tests.
+// If useParser is true the parser is used to generate the file, otherwise "sigs.k8s.io/yaml" is used.
+// The TestDevfile structure specified contains the name of the devfile and its required content.
+func (devfile *TestDevfile) CreateDevfile(useParser bool) error {
+	var err error
+
+	fileName := devfile.FileName
+	if !strings.HasSuffix(fileName, ".yaml") {
+		fileName += ".yaml"
+	}
+
+	if useParser {
+		LogInfoMessage(fmt.Sprintf("Use Parser to write devfile %s", fileName))
+		newDevfile, err := devfileData.NewDevfileData(devfile.SchemaDevFile.SchemaVersion)
+		if err != nil {
+			LogErrorMessage(fmt.Sprintf("Creating new devfile : %v", err))
+		} else {
+			newDevfile.SetSchemaVersion(devfile.SchemaDevFile.SchemaVersion)
+
+			// add the commands to new devfile
+			for _, command := range devfile.SchemaDevFile.Commands {
+				newDevfile.AddCommands(command)
+			}
+			// add components to the new devfile
+			newDevfile.AddComponents(devfile.SchemaDevFile.Components)
+
+			ctx := devfileCtx.NewDevfileCtx(fileName)
+
+			err = ctx.SetAbsPath()
+			if err != nil {
+				LogErrorMessage(fmt.Sprintf("Setting devfile path : %v", err))
+			} else {
+				devObj := parser.DevfileObj{
+					Ctx:  ctx,
+					Data: newDevfile,
+				}
+				err = devObj.WriteYamlDevfile()
+				if err != nil {
+					LogErrorMessage(fmt.Sprintf("Writing devfile : %v", err))
+				} else {
+					devfile.SchemaParsed = false
+				}
+			}
+
+		}
```

Comment:
I see a lot of redundancy here. We are creating a test devfile with `devfile *TestDevfile` which ultimately is the same as `newDevfile, err := devfileData.NewDevfileData(devfile.SchemaDevFile.SchemaVersion)` 

Why not create your test devfile with `newDevfile, err := devfileData.NewDevfileData(devfile.SchemaDevFile.SchemaVersion)` in the first place?

You are storing the test data in two different structs here at this moment

Reply:
The test is meant to test the library api and as a result it should be using every api available, the code my look cumbersome but that is just how tests are. You raise an interesting point though, there may be parts of the api I am not testing fully and I need to do a review of that - make sure I fully exercise all apis.